src/vault/client/ext/aws.clj

```diff
@@ -0,0 +1,99 @@
+(ns vault.client.ext.aws
+  (:require
+    [buddy.core.codecs :refer [bytes->str]]
+    [buddy.core.codecs.base64 :as b64]
+    [cheshire.core :as json]
+    [clj-http.client :as http]
+    [clojure.java.io :as io]
+    [envoy.core :refer [defenv]]
+    [vault.client.http :as client])
+  (:import
+    (com.amazonaws
+      DefaultRequest
+      Request
+      SignableRequest)
+    (com.amazonaws.auth
+      AWS4Signer
+      AWSCredentials
+      BasicSessionCredentials
+      DefaultAWSCredentialsProviderChain)
+    com.amazonaws.http.HttpMethodName
+    java.net.URI))
+
+
+(defenv :vault-aws-iam-role
+  "The configured vault aws auth role used to perform instance authentication.")
+
+
+(def ^:private payload "Action=GetCallerIdentity&Version=2011-06-15")
+
+
+(defn sts-get-caller-identity-request
+  "Create a GetCallerIdentity request ready for signing."
+  ^SignableRequest
+  []
+  (doto (DefaultRequest. "sts")
+    (.setEndpoint (URI. "https://sts.amazonaws.com"))
+    (.setResourcePath "/")
+    (.setHttpMethod HttpMethodName/POST)
+    (.addHeader "Content-Type" "application/x-www-form-urlencoded; charset=utf-8")
+    (.addHeader "Content-Length" "43")
+    (.setContent (-> payload (.getBytes "UTF-8") (io/input-stream)))))
+
+
+(defn credentials
+  "AWSCredentials to authenticate with STS. Will fall back to default chain including instance profile."
+  ^AWSCredentials
+  ([id secret token]
+   (BasicSessionCredentials. id secret token))
+  ([]
+   (-> (DefaultAWSCredentialsProviderChain/getInstance)
+       (.getCredentials))))
+
+
+(defn signer
+  "Signing object used to perform AWS4 signing on sts request object."
+  ^AWS4Signer
+  []
+  (doto (AWS4Signer.)
+        (.setServiceName "sts")
+        (.setRegionName "us-east-1")))
+
+
+(defn- request-parameters
+  "Takes signed aws Request object to derive parameters required by vault auth backend api call.
+  https://www.vaultproject.io/api/auth/aws/index.html#login"
+  [^Request req]
+  {"iam_http_request_method" (str (.getHttpMethod req))
+   "iam_request_url" (-> (format "%s%s"
+                                 (.getEndpoint req)
+                                 (.getResourcePath req))
+                         (b64/encode)
+                         (bytes->str))
+   "iam_request_body" (-> (.getContent req)
+                          (slurp)
+                          (b64/encode)
+                          (bytes->str))
+   "iam_request_headers" (-> (.getHeaders req)
+                             (json/encode)
+                             (b64/encode)
+                             (bytes->str))})
+
+
+(defmethod client/authenticate* :aws-iam
+  [client _ aws-ctx]
+  (let [{:keys [iam-role test-credentials]} aws-ctx
+        aws-creds ^AWSCredentials (or test-credentials (credentials))
```

Comment:
Suggestion: instead of naming this `:test-credentials` how about accepting just `:credentials`? You could support a few different ways of providing this; I'm thinking something [along these lines](https://github.com/greglook/blocks-s3/blob/20413c59ae08ab684092f2bf4545bcd1a4b3a627/src/blocks/store/s3.clj#L96-L124).

Reply:
i'm renaming the keyword for now and will make a note about supporting an alternative approach for a future change but for now for my use its YAGNI.